### DISCUSSION TITLE
Zero-trust access (with double tokens)

### DISCUSSION THREAD

**Discussion by thejan2009** [magit/forge#480] [Support] [answered]
Hello there,

the Gitlab instance I'll try to connect to is behind a zero-trust proxy. This means that in addition to the normal login/token that I have to present, I also need to get through SSO login page. In a web browser this is just an additional login step. In CLI, after logging in, I can get a short-lived access token that can be sent in a separate header, e.g. `curl -H "zt-token: <token>" -H "Authorization: Bearer <authtoken>" <forge_url>`.

My plan to solve this for `forge` is to
 - get the access token
 - inject it in every request header.
 
And I have trouble figuring out how an additional header can be added to forge requests. Would redefining `ghub-headers` of https://github.com/magit/ghub/blob/master/lisp/ghub.el#L727 work or is there another way that I missed?

**Comment by tarsius** [comment, marked as answer]
Redefining `ghub--headers` make sense. Redefining `ghub--auth` would make sense but then you would have to copy more code. Still, you should probably do that and once it works we can look into adding a hook or something.

**Reply by tarsius** [reply]
Or maybe not. Anyway, one of these two functions it is.

**Reply by thejan2009** [reply]
Well, that was fun. `tmp-access-token` fetches header key and value, then I push everything to the original function.
```elisp
(advice-add 'ghub--headers :around
                (lambda (original headers &rest r)
                  (push (tmp-access-token) headers)
                  (apply original headers r))) ```

**Reply by tarsius** [reply]
Great! Please consider adding a not about this to https://github.com/magit/forge/wiki/Tips-and-Tricks, provided of course you can include `tmp-access-token` and that is sufficiently generic others could use it with their instances.

**Reply by thejan2009** [reply]
I think `tmp-access-token` function is too involved and would make the whole thing confusing - I had to handle token expiration and two different ways of obtaining the token. I'll attach the code below, feel free to include it in Tips and Tricks if you think it would be a worthwhile addition. By the way, the zero-trust system in question is [Cloudflare Access](https://developers.cloudflare.com/cloudflare-one/tutorials/cli/)

```elisp
(defvar cf-token-expiration nil "Cloudflare Access token expiration time")
(defvar cf-token-value nil "Cloudflare Access token")
(defun get-cf-access-token ()
  ;; if the token won't expire in the next 5 seconds, return existing value
  (if (time-less-p (time-add nil 5) cf-token-expiration)
      cf-token-value
    (progn
      ;; otherwise get token
      (setq cf-token-value (shell-command-to-string "cloudflared access token -app='https://gitlab.forge'"))
      ;; if unable to do that, login and get token again
      (if (string-match-p (regexp-quote "Unable to find token for provided application") cf-token-value)
          (progn (shell-command-to-string "cloudflared access login 'https://gitlab.forge'")
                  (setq cf-token-value (shell-command-to-string "cloudflared access token -app='https://gitlab.forge'")))
        nil)
      ;; parse exp
      (setq cf-token-expiration
            (seconds-to-time
              (gethash "exp"
                      (json-parse-string
                        (base64-decode-string
                        (nth 1 (split-string cf-token-value "\\.")) t)))))
    cf-token-value)))

;; inject "cf-access-token" in ghub--headers function
(advice-add 'ghub--headers :around
            (lambda (original-fun headers host &rest r)
              (if (string-match-p (regexp-quote "gitlab.forge") host)
                  (push (cons "cf-access-token" (get-cf-access-token)) headers)
                nil)
              (apply original-fun headers host r)))

```
